### ISSUE TITLE
Allow blocks to support multiple block styles

### ISSUE THREAD

**Issue by ChemicalSailor** [WordPress/gutenberg#14598]
Related: https://github.com/WordPress/gutenberg/issues/7551

I'd like to propose that blocks can optionally support multiple block styles at the same time. Currently plugins/themes can add other styles, but the user has to select Style1 OR Style2 OR Style3 etc. I'd like to see the possibility that a user can select Style1 AND Style2 AND Style3 etc.

Take for example the table block. Currently it supports a basic no style default or striped rows. A plugin could add a styles for a header row and for a header column. The user should be able to choose to have a combination of header row, column and striped rows.

This wouldn't be appropriate for all blocks, so there would have to be some kind of 'supports' definition in the block registration.

(The table block may not be the most appropriate example because the table markup will need modification instead of simply adding more classes; this is something the plugin would have to sort out)

The alternative is the plugin adding a separate panel in the sidebar most likely using checkboxes to turn on/off header rows etc, but IMO this would be a worse UX than using block style variations.

**Comment by QuietNoise** [comment]
I support this feature request wholeheartedly. Much needed feature especially for static blocks. Right now we end-up adding tonnes of block styles to cover combination of cascading styles.
TinyMCE (styles selection control) does it already if you look for inspiration how to tackle this.

**Comment by ChemicalSailor** [comment]
> Much needed feature especially for static blocks. Right now we end-up adding tonnes of block styles to cover combination of cascading styles.

@QuietNoise can you add some more detail on your use case? I'm sure it will be valuable info for anyone who looks at implementing this.

I'm not sure my example with the table block is actually that relevant, especially now there is built in support for table headers, and fundamentally it's changing the table structure, not just the style. Although it would still be nice to have the preview UX like you get with block styles. @mapk it would good to get your thoughts as you have been working on the table block.

**Comment by QuietNoise** [comment]
Use cases are any scenarios where you need to combine different styles but also allowing admins to apply each style individually. 
Simple example would be to manipulate how H1 tag is styled. For example it might be in in 
- in regular font, bold font, light or black font,
- it might be with regular letter spacing, with extra letter spacing
- it might have shadow, border, background image, 
or any combination of the above. This gives us 4 x 2 x 3 = 24 style options to accommodate all combinations. Quite a hindrance to ask admins to scroll through that. Add one more option would make combinations go to 48 and so on.

Alternative would to add custom controls to Heading control but as far as my experience goes it would introduce the block validation nightmare for all existing posts (due to it being a static block). Not good either.

**Comment by mapk** [comment]
> @mapk it would good to get your thoughts as you have been working on the table block.

Thanks for the ping, @ChemicalSailor. It would be great to have various styles as settings that can be toggled on/off to get the right combination. We're working on adding more base styles to blocks and keeping here: https://github.com/WordPress/gutenberg/issues/15899, https://github.com/WordPress/gutenberg/issues/8171.

It's taking time to figure out the correct amount features and which ones we should offer as individual styles. With Gutenberg, we're going to lean more toward _less options_ right now. We'll offer style variations to help supplement that. A long list of style options can quickly become overwhelming. Ultimately we _should_ be relying on the themes to provide the variety of style options.

**Comment by mwhiteley16** [comment]
I would definitely support allowing for multiple styles assigned to one block. The examples provided by QuiteNoise are perfect. I, for example, have a site that has an alternate body style font, so I have added that as a block style that I can apply to paragraphs. As the same time, I have a block styles that assign a max-width for paragraphs, it would be nice to assign both styles at the same time so I have max-width and alternate body style, without the need to create another block style for max with AND alternate body style.

**Comment by strarsis** [comment]
Yes, adding stylistic options to a block quickly results in very large numbers of block styles for all combinations.
My current workaround is to let the theme add block options that just toggle some `has-` CSS classes.
This requires to add JavaScript/React-code that also break sometimes. Combinable block styles are much better because the theme designer can just get the job done and doesn't have to add custom UI stuff.

**Comment by ragudesign** [comment]
I agree with this solution. I added my own custom testimonial block and i need to add so many style variants as i can only pick one style. @niaccurshi suggested a very good solution, this way we can create a style group, let's say 'Border Style' and from there you can select either a dotted border or straight line. And add another style group for 'Background Style' where the user can select a pre-defined background. At the moment you need to create a style with let's say dotted border with blue background, dotted border with green background etc.

**Comment by 4mation-roger** [comment]
+1
ran up against this limitation today

**Comment by reedcodes** [comment]
I also have a need for this. I am adding block styles through my theme, using options I don't want to include in my other theme, and my block has two border styles and six background styles. It would be much easier to code and maintain if these could be grouped into two different block style groups, rather than adding one block style declaration with 12 block styles.

**Comment by strarsis** [comment]
Yes, the cascading nature of CSS would be something very useful here!

**Comment by davouid** [comment]
This what we need, for example the button:
I have options like simple, tiny, icon etc... that are all cumulative of course.
I add those via custom class. If i could provide the editor with a UI to choose these classes! It would be a dream comes true.

**Comment by strarsis** [comment]
+1! HTML and CSS already support multiple classes per element. The UI for block style selection has to be adjusted from being radio-button like to checkbox-button like and toggle the associated `is-style-` classes.

**Comment by richrd** [comment]
If there was someone to write a PR for this, would it potentially be accepted as a feature?

**Comment by strarsis** [comment]
@richard: Yes! This would be great. I am not good enough with React + Gutenberg yet to write something like this.

**Comment by skorasaurus** [comment]
+1 not so much for myself this would be a big roadblock for anyone who wishes to atomic CSS in Gutenberg.

**Comment by richrd** [comment]
I just setup a development environment and I'm taking a look at how this could/should be done. I assume this is where most of the changes need to happen https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-styles/index.js

Anyone got any pointers on what to look out for as I'm not yet familiar with the gutenberg code base?
What things might break and need fixing if a block has multiple active styles?

**Comment by strarsis** [comment]
My 2 cts: There is a preview selection UI thing used for the block style selection. Does that UI control already supports the selection of multiple items?

**Comment by richrd** [comment]
@strarsis Excellent point! I bet it doesn't but I'm not sure it would even be logical that way. I think each style preview should only show what that specific style does to the block. That way they only represent what they do on their own. The combined effect of multiple block styles should be shown in the editor but IMO not in the block style preview.

**Comment by richrd** [comment]
I implemented support for multiple styles in #21714 . It doesn't separate styles into groups, simply just allows selecting multiple block styles. I think creating style groups could be a separate enhancement but that involves a lot more work and likely involves a breaking change for the style registration API.

**Comment by carolinan** [comment]
How do we indicate styles that will not work together?
For example if there is more than one bordet style, or as in TwentyTwenty, more than one style for the separator block.

All selected styles have a check mark, but only one style is showing.

**Comment by richrd** [comment]
That's a good question. I'm not sure what the solution to that should ultimately be. Many of such conflicting styles should be self evident, such as multiple border styles, but other combinations that don't work together can be less obvious.

Still allowing many conflicting styles to be selected could be problematic. On the other hand we have no control over what sort of block styles various themes and plugins will add, so we can't reliably prevent them from conflicting. The grouping mechanism that @niaccurshi suggested would be nice but would only work within one theme/plugin, not accross all of them.

Any thoughts @mapk ?

**Comment by strarsis** [comment]
@richrd: Well the user can already select bad foreground and background color-combinations and Gutenberg doesn't prevent the user from doing that. Instead, it just shows a notice/warning.

**Comment by ChemicalSailor** [comment]
A lot has changed since I first created this issue. I think there is still merit in this idea, I really like the way block styles give you a little preview of what the block will look like, and the simplicity of it just adding a class to the block actually gives a very easy entry point to customising a block.

However, I'm leaning towards block styles becoming more of 'block presets' feature - something that sets one _or more_ options in one go, kind of like the patterns API but in the local scope of the block (the presets could even be grouped - light, dark etc). A block style is really a combination of one or more features, so I think each feature should have an individual control (e.g a toggle) with associated styling where appropriate. There is some discussion in https://github.com/WordPress/gutenberg/issues/19659 which I think is relevant here.

Thinking out loud - perhaps there is scope to have both block styles and block presets as two different tabs of the same panel. I suspect it wouldn't make for very good UX, but thought I'd throw it out for discussion.

A rough, not very well thought out, way this might work:

- There is an API for registering block features, similar to the current block styles API
- The developer provides a feature name, the type of control for automatic rendering (binary, slider, colour etc), the permissible values, and a callback for implementing
- Gutenberg can provide some ready-made callbacks, such as for adding a class to the block
- More advanced features would be implemented by the developer using components, but the feature still has to be registered with it's permissible values.
- There is an API for registering block presets, possibly JSON-like which specifies a name for the preset, the features it uses and their values.
- The presets can be part of the block definition or added by a theme, or even saved by the user.

I'm sure a lot of this overlaps with patterns, reusable blocks, templates, global styles and whatever else is being cooked up at the moment. Now it's late, so I will stop whittering on and see what other peoples thoughts are!

**Comment by mapk** [comment]
We now have block variations which allow for differently styled blocks to be considered and added to the Inserter. I don't think this perfectly solves this issue but worth taking a look at.

https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/block-api/block-registration.md#variations-optional

**Comment by strarsis** [comment]
@mapk: Is this already supported in the latest Gutenberg plugin? 
Can those variations also be registered using the WordPress PHP API?

**Comment by mapk** [comment]
@strarsis yes it's in the plugin. I'm not sure on the other question.

@gziolo, can you help answer this above?

**Comment by gziolo** [comment]
> Is this already supported in the latest Gutenberg plugin?

It's both in the plugin and WordPress core starting from 5.4 release.

> Can those variations also be registered using the WordPress PHP API?

It's all JavaScript-based for now. There was a dev note posted on WordPress.org blog: https://make.wordpress.org/core/2020/02/27/introduce-block-variations-api/. You can define block variations as part of the `registerBlockType` call or use it separately with [wp.blockregisterBlockVariation](https://developer.wordpress.org/block-editor/packages/packages-blocks/#registerBlockVariation) function.

**Comment by richrd** [comment]
Is there currently an easy way to define styling options for blocks that the editor can choose and allow the editor to apply multiple styles/classes to a block at once?

**Comment by Snadex** [comment]
+1 This is a must-have feature for theme authors, including the grouping functionality. 

The whole Gutenberg follows the Atomic design principles and it should go down to the css classes selection, which happens through BLock Styles. 

Currently, this is solved by writing 2-3 custom classes in the Advanced tab, which iss far from user friendly and there is no way for the user to know what are the available classes. 
If block styles go in another direction, such as Block Preset where it holds a bunch of things in it,
then at least there must be some simple way to register available CSS classes for a particular block, not just have a blank input field.

**Comment by Intervik** [comment]
As of 2021 - limit of TinyMCE support reached  - our  development team Gutenberg-ify the backend. We are *** ? *** that Gutenberg not yet support this absolute MUST support of multiple classnames. We hacked and added checkboxes in the "Advanced box" thats adds/ removes additional classnames to the "style field" and trigger change. This should be easy to filter in array of toggle checkboxes for custom addig classnames to core - if Gutenberg cant add this in the original style handler. Stop adding more blocks to the core, and cleanup the lack of functionality first.

**Comment by QuietNoise** [comment]
@Intervik any chance you can share gist of your hack's approach?

**Comment by nickfmc** [comment]
> 
> 
> How do we indicate styles that will not work together?
> For example if there is more than one border style, or as in TwentyTwenty, more than one style for the separator block.
> 
> All selected styles have a check mark, but only one style is showing.

that's exactly what I was thinking here https://github.com/WordPress/gutenberg/issues/27384

**Comment by mrwweb** [comment]
I hope everyone can agree that this issue is a love letter to the Block Styles API.

The amount of power Block Styles put in developers' hands with only a few lines of code and no build process is possibly unparalleled. Add auto-generated previews to that, and you have an extremely developer- _**and**_ user-friendly feature. So the big request in this issue is to let developers do more with the same API (and make more like them too!).

I recently posted about [my experience making a second set of faux Block Styles](https://mrwweb.com/supporting-multiple-block-styles-in-the-wordpress-block-editor/). Here was the result, and a preview of the change I propose for Block Styles.

![Image block with two sidebar panels that both look like Block Styles. One is the default Styles panel and the other is a custom Brush Stroke settings panel with visual preview of styles](https://user-images.githubusercontent.com/1238696/137933988-6dfb0871-6012-4866-a388-8589d52bee9a.png)

https://user-images.githubusercontent.com/1238696/137935133-ddf3356d-94d8-4332-b0eb-0535774eec35.mp4

With this experience, I now have a clear proposal that accounts for backwards compatibility and hopefully wouldn't be too hard to implement.

## Proposal for "Block Style Categories"
I agree that simply allowing people to select multiple styles in the current system would end up resulting in a poor editor experience. Introducing "Block Style Categories"—just like Block Pattern Categories—is the answer.

I am not the first person to propose this. I see @LeoSeyers had a similar proposal on #33093.

I propose that a new `register_block_style_category()` function allow creating multiple categories of styles, and then only one style can be selected from each category.  Each category would get it's own panel in the editor sidebar, to allow independent collapsing and a clear division between categories.

Here's the proposed code I had in my post:

```php
/* this function does not exist, only a proposal */
register_block_style_category(
	'my-style-category',
	array( 'label' => __( 'My Style Category', 'mytextdomain' ) ),
);

/* this is a real function, but 'category' attribute is just proposed */
register_block_style(
	$block,
	[
		'name' => 'my-style',
		'label' => __( 'My Style', 'mytextdomain' ),
		'category' => 'my-style-category', // optional! defaults to a general category
	]
);
```
For both backwards compatibility and simplicity, the `category` attribute of styles would be optional, and all styles without a `category` would show up in the default "Block Styles" panel. (Again, this is all parallel to the existing Block Patterns API.) The default category would still only allow one style selection, just like any other category.

## Examples enabled by this change

This change would enable the requests people are making on this issue while also protecting editors. Here are some examples  that would work really well with this setup:

- Image: Image Masks + Brush Strokes (example from my blog post)
- Image: Border Styles + Drop-shadows
- Gallery/Image: Image Filters + Image Masks
- Buttons: Animations/Transitions + Hover Effects
- Lists: Layouts (Columns) + Markers
- Table: Borders + Row Backgrounds

## Concerned about too many styles?
_People are already making too many styles. This will make it possible to have _fewer_ styles in some cases and at least make having too many styles more approachable by breaking them up._

It is easy to imagine that folks might go overboard with such a system, but that is already true of the existing system and Block Patterns (and even Blocks, for that matter). What's important is that this change would provide a path to _reduce_ the number of Block Styles for many folks, as it's definitely clear that people are making 25 Block Styles when what they really have are 2 sets of 5 mutually-exclusive styles.

Further, this would help clarify the interface in cases where both a theme and a plugin have registered block styles, so long as at least one of them registered a custom category. (I would think it would be a best practice for plugins to always register at least one category.)

## Think this is redundant with other (existing and upcoming) block settings?
_Block Styles are specifically good at providing styles that aren't achievable with editor settings._

I would refer everyone to @kjellr's [presentation about block styles from WordCamp US 2021](https://youtu.be/_6-yAZRXY8k?t=23717), showing how complex (and awesome!) a single block style can be. (And as you're watching, imagine how some of the featured block styles could become more powerful and interesting by breaking them apart or putting them together to allow for new interesting combinations.) Indeed, many block styles will probably become unnecessary as the number of block settings in the editor increases, but I think that will only clarify the powerful and lasting usefulness of Block Styles.

---

I hope that all makes sense and, even more, that it piques the interest of folks who can help make this happen!

**Comment by strarsis** [comment]
@mrweb: That's nice! I also would have an use case for this.
The user can select wavy-shaped group blocks. And then further specify in what direction the wave will go and what transition (splash; droplets; etc). Having a second block style selector just for this helps a lot with keeping the number of possible combinations low.

**Comment by nickfmc** [comment]
I do love that idea of Style Groups, but I would still also love a modifiers too that could just be a set of true false toggles that are all classes that can work together no matter on or off.  Things like the WP columns block, I'd have Modifiers: no gutters/stack early/swap order on mobile , etc. these are are things that could all be added or none be added.  This is the main reason developers are currently using 3rd party plugins to replace something that is perfectly fine but we can't have our clients remember a bunch of classes.

**Comment by mrwweb** [comment]
@nickfmc Your needs make a lot of sense. They also make me think of #15077 and #19367 as possible ways to handle that.

One thing I notice about your example is that the three modifiers might be just as appropriate (or better?) as 3 toggles instead of block styles since they don't have obviously unique visual previews or even impacts when editing (except for gap). Making registration of other block settings and attributes easier and faster (especially those that just toggle a class) feels like it might be even better as a new issue. (Just thinking aloud here.)

---

@MrWeb Looks like @strarsis forgot a "w". 😃

**Comment by nickfmc** [comment]
> 
> 
> @nickfmc Your needs make a lot of sense. They also make me think of #15077 and #19367 as possible ways to handle that.
> 

Ahh yes, those would be very similar use case for sure and both would make sense for the issues I see with my clients. I was just thinking toggles to allow more of a description of what they do, giving you a ? to hover or something but ya basically just comes down to more class options.

**Comment by jasmussen** [comment]
Coming back to this one here, I have a feeling (please correct me if I'm off base) that the need for multiple block styles is related to the current behavior where each style variation also adds a CSS class you can attach your custom rules to. Paired with the interface being rather prominent at the top of the inspector, I can see this being an easy way for a dev to add convenient style shortcuts that go beyond what's baked in.

If that's the case, I think there may be an opportunity to lean into that specific use case, and augment the existing interface that exists for adding classes, the one in "Advanced":

<img width="281" alt="Screenshot 2023-01-16 at 10 42 08" src="https://user-images.githubusercontent.com/1204802/212647029-5e4cf8ba-794a-44ec-833f-f603a3aeefca.png">

* With the incoming split inspector (#40204) there's less reason for "Advanced" to be collapsed and de-emphasized
* Just like how tags have auto-complete, a similar interface could be explored that would add class "chips", and let you search for classes already applied.

I'd love to see some great designs for improving that class interface, and making it all the more ergonomic to use. There's no reason it can't be as easy to use as the style variation panel. Embracing this as the UI for it would also avoid the pitfalls of the style variation UI, where you'd be able to apply styles that were designed to be mutually exclusive, such as Solid and Outlined, as well as the race condition headache where you'd have to decide which box-shadow would win, if two styles both added them.

On a separate note, style variations were recently made user customizable (#47138). Ultimately this can lead to a cool use case where a particular applied button style can be changed across blocks by simply editing the source variation in the style UI.

**Comment by shoelaced** [comment]
Just ran into a scenario where this would be extremely helpful and I think @mrwweb's solution is a good one. Given the concern about a cluttered panel experience, I wanted to add that unless I'm missing something and it's possible already, it would be great to include a `styles_per_row` attribute to the style group registration. In my particular case, I could set it to a single row of 5, since the style involves small icons and the preview gives a larger view anyway, which would be much less cluttered than the 5 giant tiles at 2 per row that take up helf the panel as I'm getting now.

**Comment by mrwweb** [comment]
> I have a feeling (please correct me if I'm off base) that the need for multiple block styles is related to the current behavior where each style variation also adds a CSS class you can attach your custom rules to. Paired with the interface being rather prominent at the top of the inspector, I can see this being an easy way for a dev to add convenient style shortcuts that go beyond what's baked in.

I'd put the emphasis on "prominent" and "convenient". In my opinion, the two "killer features" of the styles interface are:

- Discoverability (closely related to "prominence" but slightly different)
- Autopreviews (some of that value has been diminished since the interface was changed to only show previews on hover)

That discoverability piece is what I would worry about with the Custom CSS Class path, but I do wonder about something like the classic "frequently used tags" type interface except controllable by developers. Even then though, I wonder if something like a toggle-able checkbox would be easier to use than an "add/remove" interface, since it's easier for users to test how something looks with a toggle rather than adding something to a box above, moving the mouse, and removing it.

**Comment by cbirdsong** [comment]
The idea of a developer-controlled list of CSS classes is a potentially interesting feature that would be worth exploring, but I agree that it isn’t really equivalent to the block style system as it exists today or the expanded functionality originally proposed in this issue. The “Advanced” panel in the sidebar does not seem like the place for it or something similar.

**Comment by nickmatternow** [comment]
there are also plugins that take care of developer classes for gutenberg now too if someone needs this functionality

https://dplugins.com/downloads/plain-classes-gutenberg/

**Comment by bongojules** [comment]
Just to say I'd also massively appreciate this feature - Especially @mrwweb's proposed implementation of "[Block Style Categories](https://github.com/WordPress/gutenberg/issues/14598#issuecomment-946823689)" 

Will be super useful for sure.

**Comment by hanneslsm** [comment]
Can we please get some traction into this issue again?
It would open up immense possibilities, especially if it would be possible to have different categories (e.g. Search block: Category 1: With Label, Category 2: Button inside, Button outside, etc. https://github.com/WordPress/gutenberg/issues/51617#issuecomment-1596023326)

**Comment by kauaicreative** [comment]
Toggle style variations on and off e.g. checkbox - not radio.

**Comment by justintadlock** [comment]
I've given a lot of thought to this recently. There are definite problems with just allowing any block style to be combined. These, particularly more complex styles, will often clash and create a poor user experience.

At the same time, there's an obvious need for theme authors to be able to add multiple CSS classes to a block and present them to users through an easy-to-use UI.

And that is 100% already possible. The problem is that there's no documentation around doing that. That's where I believe the biggest change needs to happen before we explore changing the core block styles system.

I plan on running a tutorial series on the WordPress Developer Blog showing how to do this. Here's the discussion link: https://github.com/WordPress/developer-blog-content/discussions/130

If you have feedback on multiple block styles/classes that you want to add to that discussion, please do so. My plan is to begin writing the guide next week.

**Comment by strarsis** [comment]
Naturally, CSS classes can also be combined. Sure, blocks and block styles are a higher level abstraction than CSS classes, but both allow combining things - and also allowing the theme developer to add exceptions or exclusions for specific block style combinations.

**Comment by cbirdsong** [comment]
The custom controls you have in that issue are definitely worth writing up as a guide, but it seems like there's still space for a simple way to have multiple variations on block styles. Anything involving adding custom controls and/or a build process is a massive step up in complexity from `register_block_style`/ `registerBlockStyle` and would also be more likely to break with future editor updates.

**Comment by acketon** [comment]
> I've given a lot of thought to this recently. There are definite problems with just allowing any block style to be combined. These, particularly more complex styles, will often clash and create a poor user experience.
> 
> At the same time, there's an obvious need for theme authors to be able to add multiple CSS classes to a block and present them to users through an easy-to-use UI.
> 
> And that is 100% already possible. The problem is that there's no documentation around doing that. That's where I believe the biggest change needs to happen before we explore changing the core block styles system.
> 
> I plan on running a tutorial series on the WordPress Developer Blog showing how to do this. Here's the discussion link: [WordPress/developer-blog-content#130](https://github.com/WordPress/developer-blog-content/discussions/130)
> 
> If you have feedback on multiple block styles/classes that you want to add to that discussion, please do so. My plan is to begin writing the guide next week.

@justintadlock I look forward to your documentation with great interest. That documentation will be very welcomed for a project I'm working on this year. I think that adding new controls to blocks from a theme is a good option in the short term, but I still agree with others here that Block Styles should be expanded to support multiple active styles at the same time. Block Styles are so simple and accessible for Theme Developers that may not be comfortable with more advanced modifications to a block. Being able to think of block styles in a more component based system approach instead of the mess it can become now would also really help with the bloat and CSS issues faced on the theme side. 

Like @cbirdsong mentions above, I too worry about custom built components breaking between WP upgrades. I deal with that now for our existing Block plugins and it is a constant issue for us. I'd prefer to avoid that and leverage Block Styles as outlined earlier in the comments by @mrwweb and others by supporting categories or some sort of flagging system for each style to denote if it can be mixed with other styles or what styles it can be mixed with.

**Comment by tresorama** [comment]
Having multiple block styles slot will make developer life easier.
They let use Gutenberg Blocks as React/Vue/Svelte component, because "Block Style" can be seen as component props, that simply add a CSS class to the element.

So dev can create CSS classes system using a BEM approach and with not much code/time we will have a native user friendly way to add "variants" to component.

It also remove the complexity to use React just to expose a "radio/select", and gain the auto visual preview.

**Comment by zkingdesign** [comment]
Really looking forward to this, it's a natural step in the right direction, I've came across needing this more and more while developing themes recently.

**Comment by oglekler** [comment]
I've stumbled upon the same need. Block Partners and block Organizers are pretty much the same. So, I want to make one as a variation from another but limit each version to one per page, but right now, if multiple is false, I cannot add Organizers if I have Partners added. I need them separately due to the IDs (anchors) I am using, and I don't want to allow users to fill them in themselves, IDs will be missed or messed up.

**Comment by MadtownLems** [comment]
Styles being limited to just a single Style per block continues to be a major limitation for us. A combination of @mrwweb's Style Categories AND the ability to select multiple, even within the same category, would open up so many doors for developers.

I do understand the concern about how some styles might clash with others, and that could happen - but that's on the content provider at that point.  Here are some ideas to help mitigate that:

1) When multiple Styles are selected, display a message: "Selecting multiple styles may lead to unintended displays. Be sure to check your content looks as expected."

and/or

2) When registering Styles, indicate whether or not it can be selected along with other Styles, or can only be selected by itself (and enforce that in the UI).

**Comment by mrwweb** [comment]
It's been a long time since this issue first opened and things have evolved quite a bit since then. While it's very clear that people would still do really cool things with multiple styles and style categories, my own thinking has evolved toward seeing custom Block settings as a better solution in most cases. The problem then becomes:

1. Fear of backwards compatibility breaking and
2. The comparative difficulty in registering a block style vs. adding custom UI that controls a block class

The block editor components are much more stable now than when this issue first was opened (March 2019!), reducing the risks of custom controls breaking. But that still leaves the difficulty of registering which feels like a problem worth solving. I don't know the best path forward, but it might look something like extending the `attributes` properties in `block.json` to allow for declaring an expected UI and toggle-able classname (which could also be filtered with PHP).

**Comment by JiveDig** [comment]
It's worth noting that 10Up has a useful helper component for this that they are calling Block Extensions. https://gutenberg.10up.com/reference/Blocks/block-extensions/

**Comment by shoelaced** [comment]
> I've given a lot of thought to this recently. There are definite problems with just allowing any block style to be combined. These, particularly more complex styles, will often clash and create a poor user experience.
> 

Definitely agree with this but it needn't necessarily be _any_ combination of block styles, rather _certain_ block styles which are allowed to be combined only with certain others, at least that's what I would assume such a feature would include.

I've only skimmed the comments I've missed above so perhaps this has been proposed, but would it not be possible to enable something like the style categories mentioned earlier, while also enabling either unrestricted combinations or only specific combinations? Like an array that could be added to the style registration that lists which other styles a style may be combined with? In the UI, incompatible styles could be disabled and greyed out if a restricted style is chosen.

Whether users are able to mess up a design, whether styles clash, etc., comes down to context, and it seems best for developers to take on that responsibility. With more than two categories, chaos would still be possible, but chaos is already possible in Gutenberg. I appreciate that users need guardrails but from a developer view, it seems more annoying than helpful for the system to micromanage what I can and cannot allow simply because it doesn't trust that I'm not going to build a chaotic experience for my users.

In any case, I've run into enough valid use cases that I'd love to have this option.

**Comment by MadtownLems** [comment]
Great comment, @mrwweb - I agree with all of it.

> The comparative difficulty in registering a block style vs. adding custom UI that controls a block class

This is so important. Adding a block style takes less than a minute. Adding a custom UI that controls a block class takes... I don't know, because it's such a massive project I haven't even attempted to tackle it yet! (but if it still requires all this 😲 https://developer.wordpress.org/news/2024/08/how-to-extend-a-wordpress-block/#adding-custom-functionality )

I think a big part of how far to go with Block Styles hinges on if there are any plans to make it easier to add custom controls. In an ideal world, core would have some kind of Advanced Custom Fields like API for blocks. That is to say, I could define various controls for a block with simple JSON:

{
  "field_group": {
    "name": "responsive-options",
    "label": "Responsive Options",
    "location": "core/columns",
    "fields": [
      {
        "name": "reverse-on-mobile",
        "label": "Reverse on Mobile",
        "type": "toggle",
        "default": false
      }
    ]
  }
}

Frankly, I'm only encouraging the expansion and continued development of Styles because I don't see this kind of thing happening, so am trying to figure out how to use existing tools to support our needs.

**Comment by retlehs** [comment]
Here's my workaround in the meantime in less than 200 LoC:

- PHP-register your styles as usual with `register_block_style()` — those classes still drive your CSS and front-end render, this only changes the editor UI
- Mirror them into a `STYLE_GROUPS` config (per-block, organized under named groups like "Typography", "Decoration") and a `STYLE_LABELS` map
- Mark a group `exclusive: true` for radio behavior (toggling one auto-deselects siblings) or `false` for free stacking

https://gist.github.com/retlehs/f543628dc2b37a0ec1edbd6740488547

**Comment by mrwweb** [comment]
Just came across this [new plugin that adds support for multiple block styles](https://wordpress.org/plugins/multi-block-styles/).
